network/rcmgr.go

```diff
@@ -0,0 +1,313 @@
+//    a Scope method.
+//  - Services typically center around streams, where the programmer can attach streams to a
+//    particular service. They can also directly reserve memory for a service by accessing the
+//    service scope using the ResourceManager interface.
+//  - Applications that want to account for their network resource usage can reserve memory,
+//    typically using a transaction, directly in the System or a Service scope; they can also
+//    opt to use appropriate steam scopes for streams that they create or own.
+//
+// User Serviceable Parts: the user has the option to specify their own implementation of the
+// interface. We provide a canonical implementation in the go-libp2p-resource-manager package.
+// The user of that package can specify limits for the various scopes, which can be static
+// or dynamic.
+type ResourceManager interface {
+	// ViewSystem views the system wide resource scope.
+	// The system scope is the top level scope that accounts for global
+	// resource usage at all levels of the system. This scope constrains all
+	// other scopes and institutes global hard limits.
+	ViewSystem(func(ResourceScope) error) error
+
+	// ViewTransient views the transient (DMZ) resource scope.
+	// The transient scope accounts for resources that are in the process of
+	// full establishment.  For instance, a new connection prior to the
+	// handshake does not belong to any peer, but it still needs to be
+	// constrained as this opens an avenue for attacks in transient resource
+	// usage. Similarly, a stream that has not negotiated a protocol yet is
+	// constrained by the transient scope.
+	ViewTransient(func(ResourceScope) error) error
+
+	// ViewService retrieves a service-specific scope.
+	ViewService(string, func(ServiceScope) error) error
+
+	// ViewProtocol views the resource management scope for a specific protocol.
+	ViewProtocol(protocol.ID, func(ProtocolScope) error) error
+
+	// ViewPeer views the resource management scope for a specific peer.
+	ViewPeer(peer.ID, func(PeerScope) error) error
+
+	// OpenConnection creates a new connection scope not yet associated with any peer; the connection
+	// is scoped at the transient scope.
+	// The caller owns the returned scope and is responsible for calling Done in order to signify
+	// the end of th scope's span.
+	OpenConnection(dir Direction, usefd bool) (ConnManagementScope, error)
+
+	// OpenStream creates a new stream scope, initially unnegotiated.
+	// An unnegotiated stream will be initially unattached to any protocol scope
+	// and constrained by the transient scope.
+	// The caller owns the returned scope and is responsible for calling Done in order to signify
+	// the end of th scope's span.
+	OpenStream(p peer.ID, dir Direction) (StreamManagementScope, error)
+
+	// Close closes the resource manager
+	Close() error
+}
+
+// MemoryStatus is an indicator of the current level of available memory for scope reservations.
+type MemoryStatus int
```

Comment:
```suggestion
type MemoryStatus uint8
```

Reply:
does this change anything? It is not stored anywhere, so it is not using memory and it is just a register in either case.

Reply:
Not sure, but it can't hurt, can it?

Reply:
yeah, I guess not and it might lead to shorter code in x86(-64) because it wont have the prefix byte!